src/lib/storage/numa_placement_manager.cpp

```diff
@@ -0,0 +1,95 @@
+#include "numa_placement_manager.hpp"
+
+#if OPOSSUM_NUMA_SUPPORT
+#include <numa.h>
+#endif
+
+#include <boost/container/pmr/memory_resource.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <cstdio>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+#if OPOSSUM_NUMA_SUPPORT
+#include "tasks/chunk_metrics_collection_task.hpp"
+#include "tasks/migration_preparation_task.hpp"
+#endif
+
+namespace opossum {
+
+std::shared_ptr<NUMAPlacementManager> NUMAPlacementManager::_instance;
+
+const std::shared_ptr<NUMAPlacementManager>& NUMAPlacementManager::get() { return _instance; }
```

Comment:
I'm having trouble implementing this. I pushed a non-building version on https://github.com/hyrise/hyrise/compare/numa-placement-manager...numa-placement-manager-singleton. Perhaps you could help me with that.

<details>
<summary>Compiler Error</summary>

```
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp: In static member function 'static void opossum::NUMAPlacementManager::reset(std::shared_ptr<opossum::Topology>, opossum::NUMAPlacementManager::Options)':
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:28:49: error: use of deleted function 'opossum::NUMAPlacementManager& opossum::NUMAPlacementManager::operator=(const opossum::NUMAPlacementManager&)'
   get() = NUMAPlacementManager(topology, options);
                                                 ^
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note: 'opossum::NUMAPlacementManager& opossum::NUMAPlacementManager::operator=(const opossum::NUMAPlacementManager&)' is implicitly deleted because the default definition would be ill-formed:
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: no matching function for call to 'std::shared_ptr<opossum::Topology>::operator=(const std::shared_ptr<opossum::Topology>&) const'
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:296:19: note: candidate: std::shared_ptr<_Tp>& std::shared_ptr<_Tp>::operator=(const std::shared_ptr<_Tp>&) [with _Tp = opossum::Topology] <near match>
       shared_ptr& operator=(const shared_ptr&) noexcept = default;
                   ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:296:19: note:   passing 'const std::shared_ptr<opossum::Topology>*' as 'this' argument discards qualifiers
/usr/include/c++/7/bits/shared_ptr.h:300:2: note: candidate: std::shared_ptr<_Tp>::_Assignable<const std::shared_ptr<_Yp>&> std::shared_ptr<_Tp>::operator=(const std::shared_ptr<_Yp>&) [with _Yp = opossum::Topology; _Tp = opossum::Topology; std::shared_ptr<_Tp>::_Assignable<const std::shared_ptr<_Yp>&> = std::shared_ptr<opossum::Topology>&] <near match>
  operator=(const shared_ptr<_Yp>& __r) noexcept
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:300:2: note:   passing 'const std::shared_ptr<opossum::Topology>*' as 'this' argument discards qualifiers
/usr/include/c++/7/bits/shared_ptr.h:309:2: note: candidate: template<class _Yp> std::shared_ptr<_Tp>::_Assignable<std::auto_ptr<_Up> > std::shared_ptr<_Tp>::operator=(std::auto_ptr<_Up>&&) [with _Yp = _Yp; _Tp = opossum::Topology]
  operator=(auto_ptr<_Yp>&& __r)
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:309:2: note:   template argument deduction/substitution failed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note:   types 'std::auto_ptr<_Up>' and 'const std::shared_ptr<opossum::Topology>' have incompatible cv-qualifiers
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:317:7: note: candidate: std::shared_ptr<_Tp>& std::shared_ptr<_Tp>::operator=(std::shared_ptr<_Tp>&&) [with _Tp = opossum::Topology] <near match>
       operator=(shared_ptr&& __r) noexcept
       ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:317:7: note:   conversion of argument 1 would be ill-formed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: cannot bind rvalue reference of type 'std::shared_ptr<opossum::Topology>&&' to lvalue of type 'const std::shared_ptr<opossum::Topology>'
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:325:2: note: candidate: template<class _Yp> std::shared_ptr<_Tp>::_Assignable<std::shared_ptr<_Yp> > std::shared_ptr<_Tp>::operator=(std::shared_ptr<_Yp>&&) [with _Yp = _Yp; _Tp = opossum::Topology]
  operator=(shared_ptr<_Yp>&& __r) noexcept
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:325:2: note:   template argument deduction/substitution failed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note:   types 'std::shared_ptr<_Tp>' and 'const std::shared_ptr<opossum::Topology>' have incompatible cv-qualifiers
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:81:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/shared_ptr.h:333:2: note: candidate: template<class _Yp, class _Del> std::shared_ptr<_Tp>::_Assignable<std::unique_ptr<_Up, _Ep> > std::shared_ptr<_Tp>::operator=(std::unique_ptr<_Up, _Ep>&&) [with _Yp = _Yp; _Del = _Del; _Tp = opossum::Topology]
  operator=(unique_ptr<_Yp, _Del>&& __r)
  ^~~~~~~~
/usr/include/c++/7/bits/shared_ptr.h:333:2: note:   template argument deduction/substitution failed:
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: note:   types 'std::unique_ptr<_Tp, _Dp>' and 'const std::shared_ptr<opossum::Topology>' have incompatible cv-qualifiers
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: passing 'const opossum::NUMAPlacementManager::Options' as 'this' argument discards qualifiers [-fpermissive]
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:21:10: note:   in call to 'constexpr opossum::NUMAPlacementManager::Options& opossum::NUMAPlacementManager::Options::operator=(const opossum::NUMAPlacementManager::Options&)'
   struct Options {
          ^~~~~~~
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = opossum::PausableLoopThread; _Dp = std::default_delete<opossum::PausableLoopThread>]'
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:389:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^~~~~~~~
In file included from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:0:
/home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:55:25: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = opossum::PausableLoopThread; _Dp = std::default_delete<opossum::PausableLoopThread>]'
   NUMAPlacementManager& operator=(const NUMAPlacementManager&) = default;
                         ^~~~~~~~
In file included from /usr/include/c++/7/memory:80:0,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.hpp:8,
                 from /home/Norman.Rzepka/zweirise/src/lib/storage/numa_placement_manager.cpp:3:
/usr/include/c++/7/bits/unique_ptr.h:389:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^~~~~~~~
src/lib/CMakeFiles/hyrise.dir/build.make:2486: recipe for target 'src/lib/CMakeFiles/hyrise.dir/storage/numa_placement_manager.cpp.o' failed
make[2]: *** [src/lib/CMakeFiles/hyrise.dir/storage/numa_placement_manager.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:930: recipe for target 'src/lib/CMakeFiles/hyrise.dir/all' failed
make[1]: *** [src/lib/CMakeFiles/hyrise.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

</details>

Reply:
Could you share the diff that leads to this error?

Reply:
Isn't that https://github.com/hyrise/hyrise/compare/numa-placement-manager...numa-placement-manager-singleton ?

Reply:
D'oh. I only expanded the error, but missed that link. Sorry.

Reply:
If you really need that `reset` method (do you?), you will have to allow the move assignment constructor which was implicitly deleted when you set the other constructors.